README.md
Description
===========

Installs the OpenStack Dashboard service **Horizon** as part of the OpenStack reference deployment Chef for OpenStack. The http://github.com/mattray/chef-openstack-repo contains documentation for using this cookbook in the context of a full OpenStack deployment. Horizon is currently installed from packages.

http://horizon.openstack.org

Requirements
============

* Chef 0.10.0 or higher required (for Chef environment use).

Cookbooks
---------

The following cookbooks are dependencies:

* apache2
* openstack-common

Usage
=====

horizon
-------

Sets up the packages needed to run the Horizon dashboard and its dependencies.
Will be included from the `server` recipe.

apache2-server
--------------

Installs the Apache webserver and sets up an `mod_wsgi` container to run the
Horizon dashboard.  Will be included from the `server` recipe.

server
------

Sets up the Horizon dashboard and a webserver of type `['openstack']['dashboard']['server_type']`
to run it, default type is 'apache2'.

```json
"run_list": [
    "recipe[openstack-dashboard::server]"
]
```

Attributes
==========

* `openstack['dashboard']['server_type']` - Selects the type of webserver to install
* `openstack['dashboard']['db']['username']` - Username for horizon database access
* `openstack['dashboard']['server_hostname']` - Sets the ServerName in the webserver config
* `openstack['dashboard']['allowed_hosts']` - List of host/domain names we can service (default: '\[\*\]')
* `openstack['dashboard']['dash_path']` - Base path for dashboard files (document root)
* `openstack['dashboard']['wsgi_path']` - Path for wsgi dir
* `openstack['dashboard']['wsgi_socket_prefix']` - Location that will override the standard Apache runtime directory
* `openstack['dashboard']['ssl_offload']` - Set SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTOCOL', 'https') flag for offloading SSL
* `openstack['dashboard']['plugins']` - Array of plugins to include via INSTALED\_APPS
* `openstack['dashboard']['simple_ip_management']` - Boolean to enable or disable simplified floating IP address management
* `openstack['dashboard']['http_port']` - Port that httpd should listen on (default: 80)
* `openstack['dashboard']['https_port']` - Port that httpd should listen on for using ssl (default: 443)
* `openstack['dashboard']['password_autocomplete']` - Toggle browser autocompletion for login form ('on' or 'off', default: 'off')
* `openstack['dashboard']['ssl_no_verify']` - Disable SSL certificate checks (useful for self-signed certificates)
* `openstack['dashboard']['ssl_cacert']` - The CA certificate to use to verify SSL connections
* `openstack['dashboard']['misc_local_settings']` - Additions to the local_settings conf file
* `openstack['dashboard']['hash_algorithm']` - Hash algorithm to use for hashing PKI tokens

Identity
--------
* `openstack['dashboard']['identity_api_version']` - Force a specific Identity API version ('2.0' or '3', default: '2.0')
* `openstack['dashboard']['volume_api_version']` - Force a specific Cinder API version (default: '2')
* `openstack['dashboard']['keystone_multidomain_support']` - Boolean to enable multi-Domain support
* `openstack['dashboard']['keystone_default_domain']` - Default Domain if using API v3 and on a single-domain model (default: 'Default')
* `openstack['dashboard']['keystone_default_role']` - Default Keystone role assigned to project members (default: '_member_')
* `openstack['dashboard']['keystone_backend']['name']` - Keystone backend in use ('native' or 'ldap', default: 'native')
* `openstack['dashboard']['keystone_backend']['can_edit_user']` - Boolean to allow some user-related identity operations (default: true)
* `openstack['dashboard']['keystone_backend']['can_edit_group']` - Boolean to allow some group-related identity operations (default: true)
* `openstack['dashboard']['keystone_backend']['can_edit_project']` - Boolean to allow some project-related identity operations (default: true)
* `openstack['dashboard']['keystone_backend']['can_edit_domain']` - Boolean to allow some domain-related identity operations (default: true)
* `openstack['dashboard']['keystone_backend']['can_edit_role']` - Boolean to allow some role-related identity operations (default: true)

Certificate
-----------
* `openstack['dashboard']['use_ssl']` - Toggle for using ssl with dashboard (default: true)
* `openstack['dashboard']['ssl']['dir']` - Directory where ssl certs are stored on this system (default: platform dependent)
* `openstack['dashboard']['ssl']['cert']` - Name to use when creating the ssl certificate
* `openstack['dashboard']['ssl']['cert_url']` - If using an existing certificate, this is the URL to its location
* `openstack['dashboard']['ssl']['key']` - Name to use when creating the ssl key
* `openstack['dashboard']['ssl']['key_url']` - If using an existing certificate key, this is the URL to its location

By default the openstack-dashboard cookbook ships with a self-signed certificate from a fake organization.
It is possible to use a real production certificate from your organization by putting that certificate
somewhere where the cookbook can download it from then simply passing in the URL of the certificate, and its
corresponding key, using the 'cert_url' and 'key_url' attributes.

Testing
=====

Please refer to the [TESTING.md](TESTING.md) for instructions for testing the cookbook.

Berkshelf
=====

Berks will resolve version requirements and dependencies on first run and
store these in Berksfile.lock. If new cookbooks become available you can run
`berks update` to update the references in Berksfile.lock. Berksfile.lock will
be included in stable branches to provide a known good set of dependencies.
Berksfile.lock will not be included in development branches to encourage
development against the latest cookbooks.

License and Author
==================

|                      |                                                    |
|:---------------------|:---------------------------------------------------|
| **Author**           |  Justin Shepherd (<justin.shepherd@rackspace.com>) |
| **Author**           |  Jason Cannavale (<jason.cannavale@rackspace.com>) |
| **Author**           |  Ron Pedde (<ron.pedde@rackspace.com>)             |
| **Author**           |  Joseph Breu (<joseph.breu@rackspace.com>)         |
| **Author**           |  William Kelly (<william.kelly@rackspace.com>)     |
| **Author**           |  Darren Birkett (<darren.birkett@rackspace.co.uk>) |
| **Author**           |  Evan Callicoat (<evan.callicoat@rackspace.com>)   |
| **Author**           |  Jay Pipes (<jaypipes@att.com>)                    |
| **Author**           |  John Dewey (<jdewey@att.com>)                     |
| **Author**           |  Matt Ray (<matt@opscode.com>)                     |
| **Author**           |  Sean Gallagher (<sean.gallagher@att.com>)         |
| **Author**           |  Chen Zhiwei (<zhiwchen@cn.ibm.com>)               |
| **Author**           |  Jian Hua Geng (<gengjh@cn.ibm.com>)               |
| **Author**           |  Ionut Artarisi (<iartarisi@suse.cz>)              |
| **Author**           |  Eric Zhou (<iartarisi@suse.cz>)                   |
| **Author**           |  Jens Rosenboom (<j.rosenboom@x-ion.de>)           |
|                      |                                                    |
| **Copyright**        |  Copyright (c) 2012, Rackspace US, Inc.            |
| **Copyright**        |  Copyright (c) 2012-2013, AT&T Services, Inc.      |
| **Copyright**        |  Copyright (c) 2013, Opscode, Inc.                 |
| **Copyright**        |  Copyright (c) 2013-2014, IBM, Corp.               |
| **Copyright**        |  Copyright (c) 2013-2014, SUSE Linux GmbH.         |
| **Copyright**        |  Copyright (c) 2014, x-ion GmbH.                   |

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
openstack-dashboard Cookbook CHANGELOG
==============================
This file is used to list changes made in each version of the openstack-dashboard cookbook.

## 10.0.0
* Upgrading to Juno
* Upgrading berkshelf from 2.0.18 to 3.1.5
* Allow enable_filewall and enable_vpn to be configured via attributes
* Sync conf files with Juno
* Add optional section support for local_settings template
* Update local_settings from 0644 to 0640
* Fix python-ibm-db-django package polluting common package attribute
* Allow some ceitificate options to be configured
* Add sensitive flag to private key and certificate file resources
* Add hash algorithm option to local_settings
* Update user group for local_settings from root to horizon user group
* Bump Chef gem to 11.16
* Add sensitive flag to local_settings file resource
* allow override of the temporary directory used for file uploads
* Set default to use only TLS for SSL. OpenStack security note OSSN-0039
* Allow TraceEnable to be configured
* Allow volume_api_version to be configured for Horizon
* Allow webserver to be configurable
* Fix site template directory defaults for apache 2.4
* Use common specific_endpoint routines (bug 1412919)
* Fix notify when using listen_addresses
* Change default for password_autocomplete to off for better default security

## 9.1
* python_packages database client attributes have been moved to the -common cookbook
* bump berkshelf to 2.0.18 to allow Supermarket support
* fix fauxhai version for suse

## 9.0.3
* Fix LOGIN_REDIRECT_URL to be configurable on rhel

## 9.0.2
* Add support for configuring OPENSTACK_KEYSTONE_BACKEND

## 9.0.1
### Bug
* Fix openstack_keystone_default_role default
* Fix the depends cookbook version issue in metadata.rb

## 9.0.0
* Upgrade to Icehouse

## 8.1.1
### Bug
* Fix the DB2 ODBC driver issue

## 8.1.0
### Blue print
* Use the library method auth_uri_transform

## 8.0.0
### New version
* Upgrade to upstream Havana release


attributes/default.rb
# encoding: UTF-8
#
# Cookbook Name:: openstack-dashboard
# Attributes:: default
#
# Copyright 2012, AT&T, Inc.
# Copyright 2013-2014, IBM, Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Set to some text value if you want templated config files
# to contain a custom banner at the top of the written file
default['openstack']['dashboard']['custom_template_banner'] = '
# This file autogenerated by Chef
# Do not edit, changes will be overwritten
'

default['openstack']['dashboard']['server_type'] = 'apache2'

default['openstack']['dashboard']['debug'] = false

# The Keystone role used by default for users logging into the dashboard
default['openstack']['dashboard']['keystone_default_role'] = '_member_'

# This is the name of the Chef role that will install the Keystone Service API
default['openstack']['dashboard']['keystone_service_chef_role'] = 'keystone'

default['openstack']['dashboard']['server_hostname'] = nil
default['openstack']['dashboard']['use_ssl'] = true
default['openstack']['dashboard']['ssl']['cert_url'] = nil
default['openstack']['dashboard']['ssl']['key_url'] = nil
# When using a remote certificate and key, the names of the actual installed certificate
# and key in the file system are determined by the following two attributes.
# If you want the name of the installed files to match the name of the files from the URL,
# they need to be manually set below, if not the conventional horizon.* names will be used.
default['openstack']['dashboard']['ssl']['cert'] = 'horizon.pem'
default['openstack']['dashboard']['ssl']['key'] = 'horizon.key'
# Which versions of the SSL/TLS protocol will be accepted in new connections.
default['openstack']['dashboard']['ssl']['protocol'] = 'All -SSLv2 -SSLv3'

# List of hosts/domains the dashboard can serve. This should be changed, a '*'
# allows everything
default['openstack']['dashboard']['allowed_hosts'] = ['*']

default['openstack']['dashboard']['swift']['enabled'] = 'False'

default['openstack']['dashboard']['theme'] = 'default'

default['openstack']['dashboard']['apache']['sites-path'] = "#{node['apache']['dir']}/openstack-dashboard.conf"

# Allow TRACE method
#
# Set to "extended" to also reflect the request body (only for testing and
# diagnostic purposes).
#
# Set to one of:  On | Off | extended
default['openstack']['dashboard']['traceenable'] = node['apache']['traceenable']

default['openstack']['dashboard']['http_port'] = 80
default['openstack']['dashboard']['https_port'] = 443

default['openstack']['dashboard']['secret_key_content'] = nil

default['openstack']['dashboard']['ssl_no_verify'] = 'True'
default['openstack']['dashboard']['ssl_cacert'] = nil

default['openstack']['dashboard']['webroot'] = '/'

# Dashboard specific database packages
# Put common ones here and platform specific ones below.
default['openstack']['dashboard']['db_python_packages'] = {
  db2: [],
  mysql: [],
  postgresql: [],
  sqlite: []
}

# The hash algorithm to use for authentication tokens. This must match the hash
# algorithm that the identity (Keystone) server and the auth_token middleware
# are using. Allowed values are the algorithms supported by Python's hashlib
# library.
default['openstack']['dashboard']['hash_algorithm'] = 'md5'

case node['platform_family']
when 'rhel'
  default['openstack']['dashboard']['horizon_user'] = 'apache'
  default['openstack']['dashboard']['horizon_group'] = 'apache'
  default['openstack']['dashboard']['secret_key_path'] = '/usr/share/openstack-dashboard/openstack_dashboard/local/.secret_key_store'
  default['openstack']['dashboard']['ssl']['dir'] = '/etc/pki/tls'
  default['openstack']['dashboard']['local_settings_path'] = '/etc/openstack-dashboard/local_settings'
  default['openstack']['dashboard']['django_path'] = '/usr/share/openstack-dashboard'
  default['openstack']['dashboard']['login_url'] = "#{node['openstack']['dashboard']['webroot']}auth/login/"
  default['openstack']['dashboard']['logout_url'] = "#{node['openstack']['dashboard']['webroot']}auth/logout/"
  default['openstack']['dashboard']['login_redirect_url'] = node['openstack']['dashboard']['webroot']
  default['openstack']['dashboard']['db_python_packages']['db2'] = ['python-ibm-db-django']
  default['openstack']['dashboard']['platform'] = {
    'horizon_packages' => ['openstack-dashboard'],
    'memcache_python_packages' => ['python-memcached'],
    'package_overrides' => ''
  }
  default['openstack']['dashboard']['apache']['sites-path'] = "#{node["apache"]["dir"]}/sites-available/openstack-dashboard.conf"
when 'suse'
  default['openstack']['dashboard']['horizon_user'] = 'wwwrun'
  default['openstack']['dashboard']['horizon_group'] = 'www'
  default['openstack']['dashboard']['secret_key_path'] = '/srv/www/openstack-dashboard/openstack_dashboard/local/.secret_key_store'
  default['openstack']['dashboard']['ssl']['dir'] = '/etc/ssl'
  default['openstack']['dashboard']['local_settings_path'] = '/srv/www/openstack-dashboard/openstack_dashboard/local/local_settings.py'
  default['openstack']['dashboard']['django_path'] = '/srv/www/openstack-dashboard'
  default['openstack']['dashboard']['login_url'] = nil
  default['openstack']['dashboard']['logout_url'] = nil
  default['openstack']['dashboard']['login_redirect_url'] = nil
  default['openstack']['dashboard']['platform'] = {
    'horizon_packages' => ['openstack-dashboard'],
    'memcache_python_packages' => ['python-python-memcached'],
    'package_overrides' => ''
  }
  default['openstack']['dashboard']['apache']['sites-path'] = "#{node["apache"]["dir"]}/conf.d/openstack-dashboard.conf"
when 'debian'
  default['openstack']['dashboard']['horizon_user'] = 'horizon'
  default['openstack']['dashboard']['horizon_group'] = 'horizon'
  default['openstack']['dashboard']['secret_key_path'] = '/var/lib/openstack-dashboard/secret_key'
  default['openstack']['dashboard']['ssl']['dir'] = '/etc/ssl'
  default['openstack']['dashboard']['local_settings_path'] = '/etc/openstack-dashboard/local_settings.py'
  default['openstack']['dashboard']['django_path'] = '/usr/share/openstack-dashboard'
  default['openstack']['dashboard']['login_url'] = nil
  default['openstack']['dashboard']['logout_url'] = nil
  default['openstack']['dashboard']['login_redirect_url'] = nil
  default['openstack']['dashboard']['platform'] = {
    'memcache_python_packages' => ['python-memcache'],
    'package_overrides' => "-o Dpkg::Options::='--force-confold' -o Dpkg::Options::='--force-confdef'"
  }
  default['openstack']['dashboard']['platform']['horizon_packages'] = ['node-less', 'openstack-dashboard']
  default['openstack']['dashboard']['apache']['sites-path'] = "#{node["apache"]["dir"]}/sites-available/openstack-dashboard.conf"
end

default['openstack']['dashboard']['dash_path'] = "#{node['openstack']['dashboard']['django_path']}/openstack_dashboard"
default['openstack']['dashboard']['static_path'] = "#{node['openstack']['dashboard']['django_path']}/static"
default['openstack']['dashboard']['stylesheet_path'] = '/usr/share/openstack-dashboard/openstack_dashboard/templates/_stylesheets.html'
default['openstack']['dashboard']['wsgi_path'] = node['openstack']['dashboard']['dash_path'] + '/wsgi/django.wsgi'
default['openstack']['dashboard']['wsgi_socket_prefix'] = nil
default['openstack']['dashboard']['session_backend'] = 'memcached'

default['openstack']['dashboard']['ssl_offload'] = false
default['openstack']['dashboard']['plugins'] = nil

default['openstack']['dashboard']['file_upload_temp_dir'] = nil

default['openstack']['dashboard']['error_log'] = 'openstack-dashboard-error.log'
default['openstack']['dashboard']['access_log'] = 'openstack-dashboard-access.log'

default['openstack']['dashboard']['help_url'] = 'http://docs.openstack.org'

default['openstack']['dashboard']['csrf_cookie_secure'] = true
default['openstack']['dashboard']['session_cookie_secure'] = true

default['openstack']['dashboard']['keystone_multidomain_support'] = false
default['openstack']['dashboard']['api']['auth']['version'] = node['openstack']['api']['auth']['version']

case node['openstack']['dashboard']['api']['auth']['version']
when 'v2.0'
  default['openstack']['dashboard']['identity_api_version'] =  2.0
when 'v3.0'
  default['openstack']['dashboard']['identity_api_version'] = 3
end

default['openstack']['dashboard']['volume_api_version'] = 2
default['openstack']['dashboard']['keystone_default_domain'] = 'Default'
default['openstack']['dashboard']['console_type'] = 'AUTO'

default['openstack']['dashboard']['keystone_backend']['name'] = 'native'
default['openstack']['dashboard']['keystone_backend']['can_edit_user'] = true
default['openstack']['dashboard']['keystone_backend']['can_edit_group'] = true
default['openstack']['dashboard']['keystone_backend']['can_edit_project'] = true
default['openstack']['dashboard']['keystone_backend']['can_edit_domain'] = true
default['openstack']['dashboard']['keystone_backend']['can_edit_role'] = true

default['openstack']['dashboard']['log_level']['horizon'] = 'INFO'
default['openstack']['dashboard']['log_level']['openstack_dashboard'] = 'INFO'
default['openstack']['dashboard']['log_level']['novaclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['cinderclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['keystoneclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['glanceclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['neutronclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['heatclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['ceilometerclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['troveclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['swiftclient'] = 'INFO'
default['openstack']['dashboard']['log_level']['openstack_auth'] = 'INFO'
default['openstack']['dashboard']['log_level']['nose.plugins.manager'] = 'INFO'
default['openstack']['dashboard']['log_level']['django'] = 'INFO'

default['openstack']['dashboard']['password_autocomplete'] = 'off'
default['openstack']['dashboard']['simple_ip_management'] = false
default['openstack']['dashboard']['neutron']['enable_lb'] = false
default['openstack']['dashboard']['neutron']['enable_quotas'] = true
default['openstack']['dashboard']['neutron']['enable_firewall'] = false
default['openstack']['dashboard']['neutron']['enable_vpn'] = false

# Allow for misc sections to be added to the local_settings template
# For example: {
#                'CUSTOM_CONFIG_A' => {
#                  'variable1': 'value1',
#                  'variable2': 'value2'
#                }
#                'CUSTOM_CONFIG_B' => {
#                  'variable1': 'value1',
#                  'variable2': 'value2'
#                }
#              }
# will generate:
#  CUSTOM_CONFIG_A = {
#    'varable1': 'value1',
#    'varable2': 'value2',
#  }
#  CUSTOM_CONFIG_A = {
#    'varable1': 'value1',
#    'varable2': 'value2',
#  }
default['openstack']['dashboard']['misc_local_settings'] = nil
